### DISCUSSION TITLE
HMC5883L not working on GY-86/87, where do I have to place extra code?

### DISCUSSION THREAD

**Discussion by FHoevi** [iNavFlight/inav#6817] [Q&A] [answered]
Hi all,
I am using a GY-87 breakout board (MPU6050, MS5611, HMC5883L) on a MAMBA F722S stack. MS5611 is working smoothly, but HMC5883L is hiding behind MPU6050 via its auxiliary I2C bus. GY-86 is essentially the same with just a BMP180 instead of MS5611 which is not relevant here.

I need to switch MPU6050 into bypass mode. This can be done via sending three I2C commands to that device, I already tested that successfully on an Arduino:

Address    Register     Value
0x68          0x6A          0
0x68          0x37          2
0x68          0x6B          0

The address is that of MPU6050, the three registers are USER_CTRL (0x6A), INT_PIN_CFG (0x37), PWR_MGMT_1 (0x6B).

I think I need to do that before the first communication with the mag already in function hmc5883lDetect of compass_hmc5883l.c but how to achieve that in the easiest way?

Any hints or corrections are very welcome.

**Comment by FHoevi** [comment, marked as answer]
Finally I found out what needs to be done:

As already conjectured, in compass_hmc5883l.c, function hmc5883lDetect, add the following lines before the rest of the code.

// GY-87, switch MPU6050 to bypass mode to allow accessing HMC5883L
i2cWrite(I2CDEV_2, MPU6050_ADDRESS, MPU6050_RA_USER_CTRL, 0b00000000, true);
i2cWrite(I2CDEV_2, MPU6050_ADDRESS, MPU6050_RA_INT_PIN_CFG, 0b00000010, true);
i2cWrite(I2CDEV_2, MPU6050_ADDRESS, MPU6050_RA_PWR_MGMT_1, 0b00000000, true);

Now the Mag indicator in INAV configurator is blue!

**Comment by FHoevi** [comment]
I forgot to mention the necessary additions to the include and #define sections at the top of the same file:

#include "drivers/bus_i2c.h"

#define MPU6050_ADDRESS 0x68
#define MPU6050_RA_USER_CTRL 0x6A
#define MPU6050_RA_INT_PIN_CFG 0x37
#define MPU6050_RA_PWR_MGMT_1 0x6B

**Reply by JulioCesarMatias** [reply]
good job, put it as "answered"

**Comment by FHoevi** [comment]
A couple of days ago I received a personal message that this procedure does not work on any target platform. After some email ping-pong it turned out that on, e.g., a MATEKF411 one needs to do it via `I2CDEV_1` instead of `I2CDEV_2` as I had in case of my MAMBAF722 target.

Thus it is maybe a good idea first to have a look into the corresponding target.h of your choice. Somewhere in that file you will find a definition of `DEFAULT_I2C_BUS`. in case of MAMBAF722 this is set to be `BUS_I2C2`, in case of MATEKF411 it is `BUS_I2C1`.

In case of my MAMBAF722 I added in the "Magnetometer / Compass" section of target.h the following
`#define USE_MAG_GY87 // GY-87, switch MPU6050 to bypass mode to allow accessing HMC5883L`
`#define MPU6050_ADDRESS 0x68`
`#define MPU6050_RA_USER_CTRL 0x6A`
`#define MPU6050_RA_INT_PIN_CFG 0x37`
`#define MPU6050_RA_PWR_MGMT_1 0x6B`

This should be the same for any build target in the corresponding section since it just affects the additional MPU6050 / HMC5883L on GY-86 or GY-87.

Another addition is needed in this file which is target specific:
`#define GY_HMC5883L_PORT I2CDEV2`
for MAMBAF722 or
`#define GY_HMC5883L_PORT I2CDEV1`
for MATEKF411.

Finally in compass_hmc5883l.c, function hmc5883lDetect, add the following lines before the rest of the code.
`// GY-87, switch MPU6050 to bypass mode to allow accessing HMC5883L`
`i2cWrite(GY_HMC5883L_PORT, MPU6050_ADDRESS, MPU6050_RA_USER_CTRL, 0b00000000, true);`
`i2cWrite(GY_HMC5883L_PORT, MPU6050_ADDRESS, MPU6050_RA_INT_PIN_CFG, 0b00000010, true);`
`i2cWrite(GY_HMC5883L_PORT, MPU6050_ADDRESS, MPU6050_RA_PWR_MGMT_1, 0b00000000, true);`
which is valid for any target.
